### ISSUE TITLE
Compiler hangs on complex module relationship + inexistant module + overriden methods

### ISSUE THREAD

**Issue by nadako** [HaxeFoundation/haxe#3933]
Originally reported by @underscorediscovery, I reduced as much as I could from the snow library code (whew!):

A.hx
```haxe
class A {
    function f(v:B) {}
}

class AA extends A {
    override function f(v:B) {}
}
```

B.hx
```haxe
import A;
import Inexistant;

typedef B = {}
```

Main.hx
```haxe
class Main {
    static function main() {
        var a:B;
    }
}
```

`haxe -main Main -js main.js -v` prints the following and hangs hardcore:
```
Classpath : E:\Code\haxe\extraLibs/;;E:\Code\haxe\std/js/_std/;E:\Code\haxe\std/
Defines : true;js;hxcpp_api_level;haxe_ver;haxe3;dce
Parsed E:\Code\haxe\std/StdTypes.hx
Parsed E:\Code\haxe\std/js/_std/String.hx
Parsed E:\Code\haxe\std/js/_std/Array.hx
Parsed E:\Code\haxe\std/haxe/EnumTools.hx
Parsed E:\Code\haxe\std/js/_std/Std.hx
Parsed E:\Code\haxe\std/js/Boot.hx
Parsed Main.hx
Parsed E:\Code\haxe\std/Class.hx
Parsed E:\Code\haxe\std/haxe/PosInfos.hx
Parsed E:\Code\haxe\std/EnumValue.hx
Parsed E:\Code\haxe\std/Enum.hx
Typing String.charCodeAt
Parsed E:\Code\haxe\std/js/_std/HxOverrides.hx
Parsed E:\Code\haxe\std/js/_std/Date.hx
Typing HxOverrides.cca
Typing Date.toString
Typing HxOverrides.dateStr
Typing Date.now
Typing Date.fromTime
Typing Date.fromString
Typing HxOverrides.strDate
Typing HxOverrides.substr
Typing HxOverrides.indexOf
Typing HxOverrides.lastIndexOf
Typing HxOverrides.remove
Typing Array.indexOf
Typing HxOverrides.iter
Typing HxOverrides.__init__
Typing String.substr
Typing Array.insert
Typing Array.remove
Typing Array.lastIndexOf
Typing Array.copy
Typing Array.iterator
Typing haxe.EnumTools.getName
Parsed E:\Code\haxe\std/js/_std/Type.hx
Typing Type.getEnumName
Typing Type.getClass
Typing js.Boot.getClass
Typing Std.is
Typing js.Boot.__instanceof
Typing js.Boot.__interfLoop
Parsed E:\Code\haxe\std/IntIterator.hx
Typing js.Boot.__isNativeObj
Typing js.Boot.__nativeClassName
Typing js.Boot.__toStr
Typing js.Boot.__resolveNativeClass
Typing IntIterator.new
Typing IntIterator.hasNext
Typing IntIterator.next
Typing Type.getEnum
Typing Type.getSuperClass
Typing Type.getClassName
Typing Type.resolveClass
Typing js.Boot.isClass
Typing Type.resolveEnum
Typing js.Boot.isEnum
Typing Type.createInstance
Typing Type.createEmptyInstance
Typing Type.createEnum
Parsed E:\Code\haxe\std/js/_std/Reflect.hx
Parsed E:\Code\haxe\std/haxe/Constraints.hx
Typing Reflect.field
Typing Reflect.isFunction
Typing Reflect.callMethod
Typing Reflect.hasField
Typing Reflect.setField
Typing Reflect.getProperty
Typing Reflect.setProperty
Typing Reflect.fields
Typing Reflect.compare
Typing Reflect.compareMethods
Typing Reflect.isObject
Typing Reflect.isEnumValue
Typing Reflect.deleteField
Typing Reflect.copy
Typing Reflect.makeVarArgs
Typing Type.createEnumIndex
Typing Type.getInstanceFields
Typing Type.getClassFields
Typing Type.getEnumConstructs
Typing Type.typeof
Parsed E:\Code\haxe\std/js/_std/Math.hx
Typing Math.get_NEGATIVE_INFINITY
Typing Math.get_POSITIVE_INFINITY
Typing Math.get_NaN
Typing Math.ffloor
Typing Math.fceil
Typing Math.fround
Typing Math.isFinite
Typing Math.isNaN
Typing Math.__init__
Typing Type.enumEq
Typing Type.enumConstructor
Typing Type.enumParameters
Typing Type.enumIndex
Typing Type.allEnums
Typing haxe.EnumTools.createByName
Typing haxe.EnumTools.createByIndex
Typing haxe.EnumTools.createAll
Typing haxe.EnumTools.getConstructors
Typing haxe.EnumValueTools.equals
Typing haxe.EnumValueTools.getName
Typing haxe.EnumValueTools.getParameters
Typing haxe.EnumValueTools.getIndex
Typing js.Boot.__unhtml
Typing js.Boot.__trace
Typing js.Boot.__string_rec
Typing js.Boot.__clear_trace
Typing js.Boot.__cast
Typing Std.string
Typing Std.instance
Typing Std.int
Typing Std.parseInt
Typing Std.parseFloat
Typing Std.random
Typing Std.__init__
Typing Main.main
Parsed B.hx
Parsed A.hx
Typing AA.f
Typing A.f
```

**Comment by nadako** [comment]
note that changing anything in this example seems "fixes" hangs...

**Comment by underscorediscovery** [comment]
finding obscure bugs is my specialty :game_die:

**Comment by Simn** [comment]
These bugs are awesome, they usually take someone three hours to track down and then it's a one line fix. ;)

Here's the cdebug diff:

```
1668,1677c1668,1678
<   out 		ERROR Class not found : Inexistant
<   out 	run build-class (A.AA)
<   out 		run build-class (A)
<   out 			add type-field (A)
<   out 			init_class_done A
<   out 			add type-field (A:f)
<   out 		add force (A.AA)
<   out 		add type-field (A.AA)
<   out 		init_class_done AA
<   out 		add type-field (A.AA:f)
---
>   out 		flush build-class(load_module)
>   out 			run build-class (A.AA)
>   out 				run build-class (A)
>   out 					add type-field (A)
>   out 					init_class_done A
>   out 					add type-field (A:f)
>   out 				add force (A.AA)
>   out 				add type-field (A.AA)
>   out 				init_class_done AA
>   out 				add type-field (A.AA:f)
>   out 		flush-done
1680a1682,1699
>   out 	run force (Math)
>   out 	run force (Reflect)
>   out 	run force (Type)
>   out 	run force (StringBuf)
>   out 	run force (haxe.io.BytesInput)
>   out 	run force (haxe.Int32.Int32_Impl_)
>   out 		run @:to (haxe.Int32.Int32_Impl_)
>   out 	run force (haxe.Int32)
>   out 	run force (haxe.Int32)
>   out 	run force (haxe.Int64)
>   out 	run force (haxe.Int64)
>   out 	run force (String)
>   out 	run force (Array)
>   out 	run force (haxe.EnumTools.EnumValueTools)
>   out 	run force (Std)
>   out 	run force (Std)
>   out 	run force (Std)
>   out flush-done
```

@ncannasse: Does that tell you anything?

**Comment by Simn** [comment]
Sorry, I meant the diff between the `import Inexistant;` being there (hangs) or not (doesn't hang).

**Comment by Simn** [comment]
Dear diary,

I have established that we are creating a type loop by linking a typedef to itself. While checking the `valid_redefinition` of AA.f we unify its argument with the argument of A.f. At this time both arguments are typed as the typedef B, but its linked type is still a monomorph. This causes a recursive link when we bind the monomorph to B with B remaining linked to that same monomorph.

I am currently considering detecting the unification of equal TTypes and only unifying the type parameters in that case. However, there are some doubts in the back of my head because I am unsure if this addresses the cause or a symptom. It remains a mystery why the presence of an import to an unknown type affects this at all.

**Comment by Simn** [comment]
@ncannasse: This is not a regression and hangs even in 2.10. I'm reluctant to make changes to unify at this point, so we may want to leave this for 3.3 unless you can spot what the problem might be.

**Comment by ncannasse** [comment]
@Simn I'm focusing on getting these damn bytes done correctly, I'll take time after that to look at the remaining issues depending on how much time left I have before the release.

**Comment by Simn** [comment]
No problem, it's not very important. Let's check it after 3.2 then.

**Comment by PeyTy** [comment]
```haxe
class A extends B {}
class B extends A {}
```
hangs, eats RAM

**Comment by Simn** [comment]
That must be a regression, I'm sure this was properly erroring before. Could someone check with 3.1.3?

**Comment by Simn** [comment]
Note that my fix for #4671 did not address this.

**Comment by ncannasse** [comment]
My first try was not good enough :)
